citadel/indico_citadel/util.py

```diff
@@ -0,0 +1,32 @@
+# This file is part of the Indico plugins.
+# Copyright (C) 2002 - 2021 CERN
+#
+# The Indico plugins are free software; you can redistribute
+# them and/or modify them under the terms of the MIT License;
+# see the LICENSE file for more details.
+
+import asyncio
+from concurrent.futures.thread import ThreadPoolExecutor
+from functools import wraps
+
+from flask import current_app
+
+
+def parallelize(func, entries, batch_size=200):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        loop = asyncio.get_event_loop()
+        executor = ThreadPoolExecutor()
+        tasks = []
+        for idx, entry in enumerate(entries):
+            def run(app, *_args, **_kwargs):
+                with app.app_context():
+                    return func(*_args, **_kwargs)
+            tasks.append(loop.run_in_executor(
+                executor, run, current_app._get_current_object(), entry, entries, *args, **kwargs
+            ))
+            if len(tasks) >= batch_size or idx + 1 == len(entries):
```

Comment:
I think `or idx + 1 == len(entries)` works fine, but an alternative would be having this after the loop:

```py
if tasks:
    loop.run_until_complete(asyncio.gather(*tasks))
```

I don't have a strong opinion, but I think it may be nice to add a unit test to make sure we never forget to execute the last (partial) set of tasks. For example,. if the `+ 1 ` was wrong there's a good chance nobody would notice that

Reply:
Sure, sounds good. I didn't go for that alternative since I'm not a fan of repeating the same break condition twice.